Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -43,132 +42,149 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
         /// </summary>
-        private bool ShouldBreakExecution(ProcessPendingStorageContext context)
+        private void PrepareNextBlockFromPendingStorage(ProcessPendingStorageContext context)
         {
-            if (context.NextChainedBlock == null)
+            var blockIsInPendingStorage = this.BlockStoreLoop.PendingStorage.TryRemove(context.NextChainedBlock.HashBlock, out context.PendingBlockPairToStore);
+            if (blockIsInPendingStorage)
             {
-                this.logger.LogTrace("{0}:'{1}/{2}' [NEXTBLOCKNOTINPENDING_NULL]", nameof(context.NextChainedBlock), context.NextChainedBlock?.HashBlock, context.NextChainedBlock?.Height);
-                return true;
+                context.PendingBlockPairsToStore.Push(context.PendingBlockPairToStore);
+                context.PendingStorageBatchSize += context.PendingBlockPairToStore.Block.GetSerializedSize();
             }
+        }
 
-            if ((context.InputChainedBlock != null && (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock)))
-            {
-                this.logger.LogTrace("{0}:'{1}/{2}' [INVALID_PREVIOUS]", nameof(context.NextChainedBlock.Header.HashPrevBlock), context.NextChainedBlock.Header.HashPrevBlock, context.InputChainedBlock.HashBlock);
-                return true;
-            }
+        /// <summary>
+        /// Store missing blocks and remove them from pending blocks and set the Store's tip to <see cref="ProcessPendingStorageContext.NextChainedBlock"/>
+        /// </summary>
+        private async Task PushBlocksToRepository(ProcessPendingStorageContext context)
+        {
+            await this.BlockStoreLoop.BlockRepository.PutAsync(context.PendingBlockPairsToStore.First().ChainedBlock.HashBlock, context.PendingBlockPairsToStore.Select(b => b.Block).ToList());
 
-            if (context.NextChainedBlock.Height > this.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
-            {
-                this.logger.LogTrace("{0}:{1} / {2}:{3}' [STORE_AT_TIP]", nameof(context.NextChainedBlock), context.NextChainedBlock.Height, nameof(this.BlockStoreLoop.ChainState.HighestValidatedPoW), this.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height);
-                return true;
-            }
+            this.BlockStoreLoop.SetStoreTip(context.PendingBlockPairsToStore.First().ChainedBlock);
 
-            return false;
+            this.logger.LogTrace("({0}:{1} / {2}.{3}:{4} / {5}:{6} / {7}:{8})",
+                nameof(context.BlockStoreLoop.ChainState.IsInitialBlockDownload), context.BlockStoreLoop.ChainState.IsInitialBlockDownload,
+                nameof(context.PendingBlockPairsToStore), nameof(context.PendingBlockPairsToStore.Count),
+                context.PendingBlockPairsToStore?.Count, nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize,
+                nameof(context.BlockStoreLoop.StoreTip), context.BlockStoreLoop.StoreTip);
```

Comment:
This is wrong use of logging. If you want to log some particular state, not just method inputs, don't use method entry log - i.e. the one with `()`, instead use a descriptive log written in English. Also when using method entry logs, pls follow the same log style (i.e. not using ` / ` as delimiters).

Also consider overriding ToString of ProcessPendingStorageContext

Reply:
ToString for ProcessPendingStorageContext is a great idea. I'll implement that. Thanks!